src/engine.ml

```diff
@@ -344,38 +344,48 @@ let prf ?sids ~label ~secret ~client_random ~server_random len =
   and sha = p_hash Mirage_crypto.Hash.SHA1.(hmac, digest_size) s2 in
   Mirage_crypto.Uncommon.Cs.xor md5 sha
 
-let derive_keys session (my_key_material : State.my_key_material)
+let derive_keys ~tls_ekm session (my_key_material : State.my_key_material)
     (their_key_material : Packet.tls_data) =
   (* are we the server? *)
   let server = Cstruct.is_empty my_key_material.pre_master in
-  let ( pre_master,
-        client_random,
-        server_random,
-        client_random',
-        server_random',
-        sids ) =
-    if server then
-      ( their_key_material.pre_master,
-        their_key_material.random1,
-        my_key_material.random1,
-        their_key_material.random2,
-        my_key_material.random2,
-        (session.their_session_id, session.my_session_id) )
-    else
-      ( my_key_material.pre_master,
-        my_key_material.random1,
-        their_key_material.random1,
-        my_key_material.random2,
-        their_key_material.random2,
-        (session.my_session_id, session.their_session_id) )
-  in
-  let master_key =
-    prf ~label:"OpenVPN master secret" ~secret:pre_master ~client_random
-      ~server_random 48
-  in
+  let length = 4 * 64 in
   let keys =
-    prf ~label:"OpenVPN key expansion" ~secret:master_key
-      ~client_random:client_random' ~server_random:server_random' ~sids (4 * 64)
+    match tls_ekm with
+    | None ->
+        let ( pre_master,
+              client_random,
+              server_random,
+              client_random',
+              server_random',
+              sids ) =
+          if server then
+            ( their_key_material.pre_master,
+              their_key_material.random1,
+              my_key_material.random1,
+              their_key_material.random2,
+              my_key_material.random2,
+              (session.their_session_id, session.my_session_id) )
+          else
+            ( my_key_material.pre_master,
+              my_key_material.random1,
+              their_key_material.random1,
+              my_key_material.random2,
+              their_key_material.random2,
+              (session.my_session_id, session.their_session_id) )
+        in
+        let master_key =
+          prf ~label:"OpenVPN master secret" ~secret:pre_master ~client_random
+            ~server_random 48
+        in
+        prf ~label:"OpenVPN key expansion" ~secret:master_key
+          ~client_random:client_random' ~server_random:server_random' ~sids
+          length
```

Comment:
Can we now replace `prf` with `Tls.Handshake_crypto.pseudo_random_function`?

Reply:
great suggestion, done in https://github.com/robur-coop/miragevpn/pull/163/commits/30c263202881c97c529b857c5e1e64f8c9dd07ca